lib/device.c

```diff
@@ -38,18 +39,19 @@ static const char *path_encode(const char *path)
 	}
 
 	temp[pos] = '\0';
+
 	return temp;
 }
 
-static const char *sync_track_path(const char *base, const char *name)
+static const char *sync_track_path(const struct sync_device *d, const char *name)
 {
 	static char temp[FILENAME_MAX];
-	strncpy(temp, base, sizeof(temp) - 1);
+	strncpy(temp, d->base, sizeof(temp) - 1);
 	temp[sizeof(temp) - 1] = '\0';
 	strncat(temp, "_", sizeof(temp) - strlen(temp) - 1);
-	strncat(temp, path_encode(name), sizeof(temp) - strlen(temp) - 1);
+	strncat(temp, name, sizeof(temp) - strlen(temp) - 1);
 	strncat(temp, ".track", sizeof(temp) - strlen(temp) - 1);
-	return temp;
+	return d->path_encode_cb(temp);
```

Comment:
Is there a reason why we can't still pre-encode base and avoid re-encoding the same data over and over?

Reply:
I'd prefer not to, as the track name isn't permanently encoded either and user has not much visibility in the resulting path encoding. This also keeps the encoding of variables in the same place.

Second point is that it might be a tricky thing (or require something like "base" and "encoded_base" in struct sync_device) without modifying public api.

Reply:
Hmmpf. Thinking about this a bit more, I think this is going to get nasty quickly. The problem is that the base and the path-encoding needs to be consistent for the whole life-time of the connection, otherwise we might write some tracks to an old path and some to a new path.

Perhaps a better solution for now is to introduce a flag for the device-creation, so we could simply opt out of the encoding?

Reply:
i'd guess that that setting path encoding (or other kinds of rocket io callbacks) are something that a sensible user does only once? so dunno if changing callbacks later in the lifecycle makes sense in any real situation. :)

This current solution in the PR is imho quite easy for the public api by introducing only one new function. With introducing a flag it would probably mean splitting the existing functionality to at least two functions and leaving the existing sync_create_device function intact so that public API doesn't break.

Configurable path_encode function on the other hand would also leave more flexible option for the user that needs to take care of case-insensitive / case-sensitive file system woes, or something else, that cross-platform demoplayers might have.

Reply:
You're right, this is no worse than setting the IO-callbacks, they also need to be the same for the whole life-time.

I'm still going to ponder on this a bit, though. I get a bad feeling from fragile APIs... But perhaps it's the best trade-off in this case... Hmm...

Reply:
Ok, take your time. I'm also not hundred percent satisfied, from public api design perspective, with the PR's new public function althought it fulfills my need perfectly. :)

One better take on the matter could be following:
1) deprecate sync_io_cb
2) add struct sync_stdio_cb or similar that would handle the whole deal (read + write io and path mangling). maybe hiding the struct implementation details would be beneficial or forcing the user to access the structure based on some public API call which would then leave some room for future changes.